src/config/vars.rs

```diff
@@ -16,27 +21,63 @@ pub fn interpolate(input: &str, vars: &HashMap<String, String>) -> (String, Vec<
         r"(?x)
         \$\$|
         \$([[:word:].]+)|
-        \$\{([[:word:].]+)(?::-([^}]+)?)?\}",
+        \$\{([[:word:].]+)(?:(:?-|:?\?)([^}]*))?\}",
     )
     .unwrap();
+
     let interpolated = re
         .replace_all(input, |caps: &Captures<'_>| {
+            let flags = caps.get(3).map(|m| m.as_str()).unwrap_or("");
+            let def_or_err = caps.get(4).map(|m| m.as_str()).unwrap_or("");
             caps.get(1)
                 .or_else(|| caps.get(2))
                 .map(|m| m.as_str())
                 .map(|name| {
-                    vars.get(name).map(|val| val.as_str()).unwrap_or_else(|| {
-                        caps.get(3).map(|m| m.as_str()).unwrap_or_else(|| {
-                            warnings.push(format!("Unknown env var in config. name = {:?}", name));
+                    let val = vars.get(name).map(|v| v.as_str());
+                    match flags {
+                        ":-" => {
+                            if matches!(val, Some(v) if !v.is_empty()) {
+                                val.unwrap()
+                            } else {
+                                def_or_err
+                            }
+                        }
```

Comment:
It might be easier to read by moving the if-condition to the match statement, e.g.:

```rust
match flags {
    ":-" if matches!(val, Some(v) if !v.is_empty()) => val.unwrap(),
    ":-" => def_or_err

    // ...
}
```

Reply:
Ah this is exactly the kind of feedback I was looking for.
I'm still a bit "fresh" with the Rust pattern matching system and totally forgot you could do that. Great suggestion!
I will apply it later and test it. Thanks!

Reply:
I'd go a bit further and avoid the `unwrap` with something like:

```rust
                        ":?" => match val {
                            Some(v) if !v.is_empty() => v,
                            _ => {
                                errors.push(format!(
                                    "Non-empty env var required in config. name = {:?}, error = {:?}",
                                    name, def_or_err
                                ));
                                ""
                            }
                        }
```

Reply:
Or, rather, that snippet would replace the below arm for `:?` but you could do something similar for this match arm.

Reply:
Thanks for the ideas @jszwedko , will take a look at these later too.